### PR TITLE
Use new SubscribeTopics endpoint

### PR DESCRIPTION
**Pull request by neekolas**
# Use new SubscribeTopics endpoint

## Summary

This PR migrates from the `SubscribeEnvelopes` endpoint to the new `SubscribeTopics` endpoint that supports per-topic cursors and status messages. The changes include:

- **New SubscribeTopics endpoint**: Replaces `SubscribeEnvelopes` with per-topic cursor support
- **StatusAwareStream combinator**: Handles `SubscribeTopicsResponse` oneof variants, tracking subscription status and yielding envelope batches
- **Stream architecture update**: Migrates from `FlattenedStream` to `StatusAwareStream` while preserving existing `OrderedStream` and `TryExtractorStream` layers
- **Dead code removal**: Deletes `SubscribeEnvelopes`, `FlattenedStream`, and related implementations
- **Proto compatibility**: Fixes `BackupOptions` rename in device_sync proto

### New SubscribeTopics Endpoint
- Supports `TopicFilter` with individual cursors per topic
- Replaces single shared cursor with per-topic subscription filters
- Maintains builder pattern with `derive_builder`

### StatusAwareStream Combinator
- Handles `SubscribeTopicsResponse` oneof (Envelopes vs StatusUpdate)
- Tracks subscription status in shared `Arc<AtomicU8>`
- Yields only `Vec<OriginatorEnvelope>` batches
- Skips status updates and malformed responses

### Stream Migration
- Updates `XmtpMlsStreams` implementations to use per-topic cursors
- Removes dependency on lowest common cursor (LCC) calculations
- Preserves existing ordering and extraction functionality

### Cleanup
- Removes `SubscribeEnvelopesResponse` from `Paged` and `EnvelopeCollection` traits
- Deletes unused `FlattenedStream` combinator
- Updates Docker images to support new endpoint

The migration maintains backward compatibility at the trait level while modernizing the underlying subscription mechanism for better per-topic cursor management.

### CODE REVIEW

**Comment by neekolas** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3226?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3314** <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3314?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3227** <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3226** <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3226?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3226?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#3325** <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3325?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3301** <a href="https://app.graphite.com/github/pr/xmtp/libxmtp/3301?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>mergequeue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/xmtp?ref=merge-queue-instructions-in-stack-comment&prId=3226).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    type Item = Result<Vec<OriginatorEnvelope>, S::Error>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
+            match this.inner.as_mut().try_poll_next(cx) {
+                Poll::Ready(Some(Ok(response))) => {
```
Comment (insipx) [inline crates/xmtp_api_d14n/src/queries/stream/status_aware.rs]: can use the [`ready!`](https://doc.rust-lang.org/stable/std/task/macro.ready.html) macro to clean this up a bit

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    S: TryStream<Ok = SubscribeTopicsResponse>,
+{
+    type Item = Result<Vec<OriginatorEnvelope>, S::Error>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
```
Comment (insipx) [inline crates/xmtp_api_d14n/src/queries/stream/status_aware.rs]: this loop could be dangerous, it runs the possibility of starving other tasks if we never yield or if this task is always ready with a new item leading to an infinite loop.

instead, you can schedule the task to be awoken again with `cx.wake_by_ref()` in paths where we there may be more ready items (in `StatusUpdate` branch, on Started or `CatchupComplete` etc.)

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    #[xmtp_common::test(unwrap_try = true)]
+    async fn test_initial_state() {
+        let status = StreamStatus::new();
+        assert!(!status.has_started());
+        assert!(!status.catchup_complete());
+        assert_eq!(status.last_ping_ms(), 0);
+    }
+
+    #[xmtp_common::test(unwrap_try = true)]
+    async fn test_yields_envelopes_from_envelope_response() {
```
Comment (insipx) [inline crates/xmtp_api_d14n/src/queries/stream/status_aware.rs:195]: with pure `poll` implementations, it can often be preferred to test the implementation directly w/o introducing the external `tokio::` executor, that way the test can insert Pending in places where it may introduce an edge case / overall have more control over the polling execution flow.


the [`futures_test`](https://docs.rs/futures-test/latest/futures_test/) crate includes a bunch of really nice utilities for this in the form of extension traits to Future/Stream traits as well as macros. 


there are some nice examples of test like this in the `multiplexed` stream combinator: https://github.com/xmtp/libxmtp/blob/8e394a6223b982b7b4427408ca6ed2e80d7a888c/crates/xmtp_api_grpc/src/streams/multiplexed.rs#L101

As a bonus it reduces reliance on the heavy `tokio` crate for tests making them just a bit faster

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+pub fn status_aware<S>(s: S) -> (StatusAwareStream<S>, Arc<StreamStatus>) {
+    let status = Arc::new(StreamStatus::new());
+    let stream = StatusAwareStream {
+        inner: s,
+        status: Arc::clone(&status),
+    };
+    (stream, status)
+}
+
+impl<S> Stream for StatusAwareStream<S>
```
Comment (insipx) [inline crates/xmtp_api_d14n/src/queries/stream/status_aware.rs:97]: for these lower level streams its also nice to add a [`FusedStream`](https://docs.rs/futures/latest/futures/prelude/stream/trait.FusedStream.html) implementation, which avoids any potential panics resulting from weird ffi boundary runtime behavior that could result in a stream being polled after completion

File: crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+use xmtp_proto::api::{BodyError, Endpoint};
+use xmtp_proto::types::{GlobalCursor, Topic};
+use xmtp_proto::xmtp::xmtpv4::message_api::{
+    SubscribeTopicsRequest, SubscribeTopicsResponse,
+    subscribe_topics_request::TopicFilter as TopicFilterProto,
+};
+
+/// A topic paired with an optional cursor, representing a per-topic subscription filter.
+#[derive(Debug, Clone)]
+pub struct TopicFilterInput {
```
Comment (insipx) [inline crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs]: we should be able to use [TopicCursor](https://xmtp.github.io/libxmtp/xmtp_proto/types/struct.TopicCursor.html) here, which maintains a map of every `Topic` -> `GlobalCursor`

File: crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+/// Subscribe to topics with per-topic cursors.
+///
+/// Subscribe to topics with per-topic cursors, replacing the old `SubscribeEnvelopes`
+/// endpoint which only supported a single shared cursor across all topics.
+#[derive(Debug, Builder, Default, Clone)]
+#[builder(build_fn(error = "BodyError"))]
+pub struct SubscribeTopics {
+    #[builder(setter(each(name = "filter", into)), default)]
+    filters: Vec<TopicFilterInput>,
```
Comment (insipx) [inline crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs]: since `TopicCursor` implements [`FromIterator`](https://xmtp.github.io/libxmtp/xmtp_proto/types/struct.TopicCursor.html#impl-FromIterator%3C(Topic,+GlobalCursor)%3E-for-TopicCursor) for `(Topic, GlobalCursor)` `each` can be used in the same way

File: crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs (hunk 5 above)
Comment (neekolas) [reply crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs]: Response from :robot_face:  
  
`TopicCursor` requires a `GlobalCursor` for every topic, but `TopicFilterInput` supports `Option<GlobalCursor>` to handle first-time subscriptions where no cursor exists yet. I could split the filters into two collections (one using `TopicCursor` for topics with cursors, one for topics without), but that seems more complex than the current approach. Happy to change if you feel strongly though!

File: crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs (hunk 6 above)
Comment (neekolas) [reply crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs]: Same as above — `TopicCursor::FromIterator` requires `(Topic, GlobalCursor)` pairs, but our filters support `Option<GlobalCursor>` for topics without a cursor yet.

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs (hunk 1 above)
Comment (neekolas) [reply crates/xmtp_api_d14n/src/queries/stream/status_aware.rs]: Good call. Switched out

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs (hunk 4 above)
Comment (neekolas) [reply crates/xmtp_api_d14n/src/queries/stream/status_aware.rs:97]: Had Claude take a shot at it.

File: crates/xmtp_api_d14n/src/queries/stream/status_aware.rs (hunk 3 above)
Comment (neekolas) [reply crates/xmtp_api_d14n/src/queries/stream/status_aware.rs:195]: Great tip! Updated

File: crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs (hunk 5 above)
Comment (insipx) [reply crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs]: For a topic without a cursor, what happens if GlobalCursor is just set to `GlobalCursor::default`/empty? Could add a new function to Topic cursor to `add_empty` for convenience as well

Is there a proto gotcha if it's not a `null` value?

Mostly just trying to reuse as many types as possible

Comment (neekolas) [reply crates/xmtp_api_d14n/src/endpoints/d14n/subscribe_topics.rs]: Null and empty are equivalent to the server, so that’s an option
